docs/implementing.md

```diff
@@ -66,3 +73,69 @@ plain field (`foo`) will be of type `keyword`, with no nested field.
 Despite the fact that IDs are often integers in various systems, this is not
 always the case. Since we want to make it possible to map as many data sources
 to ECS as possible, we default to using the `keyword` type for IDs.
+
+## Using non-ECS fields in your event stream
+
+The goal of ECS is to define a common set of fields to make working across various
+event streams easier. Each event stream will have their own particularities, however,
+and ECS cannot account for all possibilities.
+It's therefore important that implementers feel at ease adding fields and
+top level objects in their schema that are not defined in ECS.
+This is expected.
+
+Doing so will however incur a risk of conflicting with future versions of ECS.
+Let's review some strategies to reduce the risk of conflicts happening.
+
+### Conflicts
+
+Before going into strategies, let's define what constitutes a conflict.
+
+#### Conflicts with ECS
+
+* A custom field has the same name as a new ECS field, but a different type
+  (most field types are incompatible). Here are concrete examples:
+  * `float` vs `long`
+  * `text` vs `keyword`
+  * `keyword` vs any numeric (`integer`, `long`, etc.)
+  * discrete field vs an object (e.g. a `keyword` field vs a field that contains other fields)
```

Comment:
What you mean is to remove the "discrete vs object" example?

I agree it's not as straightforward as the other ones. I wanted to include it because it'll speak very clearly to people who are about to go through the `source` field transition... But perhaps this is not necessary.

Reply:
What I mean use a specific example `keyword` vs `object` here. From the context I understand what you mean by `discrete` but for me it's not something I would directly understand.